sql/expression/literal.go

```diff
@@ -67,6 +68,11 @@ func (lit *Literal) String() string {
 	case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
 		return fmt.Sprintf("%d", v)
 	case string:
+		switch lit.fieldType.Type() {
+		// utf8 charset cannot encode binary string
+		case query.Type_VARBINARY, query.Type_BINARY:
+			return fmt.Sprintf("'0x%X'", v)
+		}
```

Comment:
I'm not convinced this is the best place for this logic... 

My concern is that hex-encoding binary data is really a UI concern, so it feels like this should be done (or at least controlled) by the client and not done across the board for all binary data. Ultimately, I think we just need to make sure that textual clients like `mysql` show the hex-encoded value, but for other clients (like a JDBC connection or something) they can still get the raw bytes, and not a hex-encoded string. 

Looking at the mysql client source code... the hex-encoding of binary data does seems to be done on the client side. There's a `--binary-as-hex` option that gets enabled by default for interactive use of `mysql` and in the source code, and then sends results to a `printAsHex` function. 

I'll ping you on Discord and we can brainstorm some other thoughts about the issue.

Reply:
Okay, chatting with Jennifer some more about this, and the customer issue only appears when using `explain` on this query with binary data in it. This makes sense now, since the output of explain has to be a string and we have to encode the value as hex here in the String method (as opposed to binary data returned from the engine that gets converted to hex on the client side for display reasons).